pkg/agent/nodeportlocal/npl_agent_init.go

```diff
@@ -58,36 +56,30 @@ func InitializeNPLAgent(kubeClient clientset.Interface, portRange, nodeName stri
 	return InitController(kubeClient, portTable, nodeName)
 }
 
-// InitController creates an Informer which watches for events from Pods running on the same Node where Antea agent is running.
+// InitController initializes the NPLController with appropriate Pod and Service Informers.
 // This function can be used independently while unit testing without using InitializeNPLAgent function.
-func InitController(kubeClient clientset.Interface, portTable *portcache.PortTable, nodeName string) (*k8s.Controller, error) {
-	c := k8s.NewNPLController(kubeClient, portTable)
-	c.RemoveNPLAnnotationFromPods()
-
+func InitController(kubeClient clientset.Interface, portTable *portcache.PortTable, nodeName string) (*k8s.NPLController, error) {
 	// Watch only the Pods which belong to the Node where the agent is running
-	fieldSelector := fields.OneTermEqualSelector("spec.nodeName", nodeName)
-	optionsModifier := func(options *metav1.ListOptions) {
-		options.FieldSelector = fieldSelector.String()
+	listOptions := func(options *metav1.ListOptions) {
+		options.FieldSelector = fields.OneTermEqualSelector("spec.nodeName", nodeName).String()
 	}
-	lw := &cache.ListWatch{
-		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
-			optionsModifier(&options)
-			return kubeClient.CoreV1().Pods(metav1.NamespaceAll).List(context.TODO(), options)
-		},
-		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
-			optionsModifier(&options)
-			return kubeClient.CoreV1().Pods(metav1.NamespaceAll).Watch(context.TODO(), options)
-		},
-	}
-
-	cacheStore, controller := cache.NewInformer(lw, &corev1.Pod{}, resyncPeriod,
-		cache.ResourceEventHandlerFuncs{
-			AddFunc:    c.EnqueueObj,
-			DeleteFunc: c.EnqueueObj,
-			UpdateFunc: func(old, cur interface{}) { c.EnqueueObj(cur) },
-		},
+	podInformer := coreinformers.NewFilteredPodInformer(
+		kubeClient,
+		metav1.NamespaceAll,
+		resyncPeriod,
+		cache.Indexers{},
+		listOptions,
 	)
-	c.CacheStore = cacheStore
-	c.Ctrl = controller
+
+	informerFactory := informers.NewSharedInformerFactory(kubeClient, resyncPeriod)
```

Comment:
This should use the global sharedInformerFactory initialized in cmd package, otherwise there will stil be duplicate connections and storages for service